### PR TITLE
Cache manager redis connection

### PR DESCRIPTION
**Pull request by atrifan**
* support passing a different redis connection provider
* support using environment variables
* support using AUTH for redis connections

### CODE REVIEW

**Review by andralungu — APPROVED** [review]
👍

File: test/perl/api-gateway/cache/cache.t [hunk 1]
```diff
@@ -241,16 +241,19 @@ X-Test: test
...
+                ",ttl_local=", key2.ttl_local,"}")
         ';
     }
 
 --- timeout: 5s
 --- request
 GET /t
 --- response_body_like eval
-['key1=nil,key2={"ttl_redis":"13","name":"Mark","ttl_local":"12"}']
+['key1=nil,key2={name=Mark,ttl_redis=13,ttl_local=12}']
```
Comment (ddragosd) [inline test/perl/api-gateway/cache/cache.t:256]: What Gateway/Openresty version returns this format ? Is it`1.12.x` ?

**Review by ddragosd — APPROVED** [review]

File: test/perl/api-gateway/cache/cache.t (hunk 1 above)
Comment (atrifan) [reply test/perl/api-gateway/cache/cache.t:256]: @ddragosd it's not the format - it's that test fail when you try to compare json with json (the key  order is not guaranted) => it's not consistent so I resorted to extracting the info into dedicated fields that's all - the format is still json

File: src/lua/api-gateway/cache/store/redisCache.lua [hunk 2]
```diff
@@ -163,20 +245,28 @@ function _M:evict(key)
...
-    local redis_host, redis_port = getRedisUpstream(REDIS_RW_UPSTREAM)
-    local ok, err = redis_rw:connect(redis_host, redis_port)
+
+    local upstreamConfig = self.redis_connection_provider.redis_rw_upstream_connection_data or REDIS_RW_UPSTREAM
+    if self.redis_connection_provider.redis_rw_upstream_connection_data and
+            type(self.redis_connection_provider.redis_rw_upstream_connection_data) == 'table' then
+        upstreamConfig = {
+            upstream = self.redis_connection_provider.redis_rw_upstream_connection_data.upstream,
+            password = os.getenv(self.redis_connection_provider.redis_rw_upstream_connection_data.password)
+        }
```
Comment (vdatcu) [inline src/lua/api-gateway/cache/store/redisCache.lua]: can we please extract as the return of a function or a module variable, so we don't duplicate?

**Review by vdatcu — APPROVED** [review]

**Comment by atrifan** [comment]
@vdatcu please check again

File: src/lua/api-gateway/cache/store/redisHashCache.lua [hunk 3]
```diff
@@ -36,6 +36,17 @@ function _M:__init(init_obj)
     throwIfFieldIsNil(init_obj)
 end
 
+---
+--- @params redisConnectionProvider
+--- @params redis_rw_upstream_connection_data
+--- @params redis_ro_upstream_connection_data
+--- Method that overrides the default redis_connection_provider
+function _M:setRedisConnectionProvider(redisConnectionProvider, redis_rw_upstream_connection_data, redis_ro_upstream_connection_data)
```
Comment (corinapurcarea) [inline src/lua/api-gateway/cache/store/redisHashCache.lua]: WDYT about moving this function in redisCache.lua?

File: src/lua/api-gateway/cache/store/redisCache.lua [hunk 4]
```diff
@@ -53,9 +70,60 @@ local function getRedisUpstream(upstream_name)
     return nil, nil
 end
 
-local cache_store_cls = require "api-gateway.cache.store"
+---
+--- @params upstream
+--- @return success, redis_instance
+--- Method that gets a redis object after connection - it tries to authenticate based on an environment variable - REDIS_PASS or
+--- REDIS_PASSWORD
+function DefaultRedisConnectionProvider:getConnection(upstream)
```
Comment (corinapurcarea) [inline src/lua/api-gateway/cache/store/redisCache.lua]: I think it would be cleaner if DefaultRedisConnectionProvider would be in its own file.

**Review by corinapurcarea — CHANGES_REQUESTED** [review]

**Comment by atrifan** [comment]
@corinapurcarea check again

**Review by corinapurcarea — APPROVED** [review]
👍

**Pull request merged**